// @progress/kendo-angular-inputs/dist/npm/slider/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function elementOffset(element) {
    var box = element.getBoundingClientRect();
    var documentElement = document.documentElement;
    return {
        top: box.top + (window.pageYOffset || documentElement.scrollTop) - (documentElement.clientTop || 0),
        left: box.left + (window.pageXOffset || documentElement.scrollLeft) - (documentElement.clientLeft || 0)
    };
}
function limitValue(value, min, max) {
    return Math.max(Math.min(value, max), min);
}
/**
 * @hidden
 */
function decreaseValueToStep(value, _a) {
    var max = _a.max, min = _a.min, smallStep = _a.smallStep;
    var stepValue = value - min;
    var result;
    if (stepValue % smallStep === 0) {
        result = stepValue - smallStep;
    }
    else {
        result = stepValue - (stepValue % smallStep);
    }
    return limitValue(result + min, min, max);
}
exports.decreaseValueToStep = decreaseValueToStep;
/**
 * @hidden
 */
function increaseValueToStep(value, _a) {
    var max = _a.max, min = _a.min, smallStep = _a.smallStep;
    var stepValue = value - min;
    var result = stepValue - (stepValue % smallStep) + smallStep;
    return limitValue(result + min, min, max);
}
exports.increaseValueToStep = increaseValueToStep;
/**
 * @hidden
 */
function snapValue(value, options) {
    var smallStep = options.smallStep, min = options.min;
    var left = decreaseValueToStep(value, options);
    var right = increaseValueToStep(value, options);
    if ((value - min) % smallStep === 0) {
        return value;
    }
    if (right - value <= (right - left) / 2) {
        return right;
    }
    return left;
}
exports.snapValue = snapValue;
/**
 * @hidden
 */
function eventValue(eventArgs, scaleElement, options) {
    var min = options.min, max = options.max, vertical = options.vertical, rtl = options.rtl;
    var trackOffset = elementOffset(scaleElement);
    var offset = vertical ? eventArgs.pageY - trackOffset.top : eventArgs.pageX - trackOffset.left;
    var scale = (max - min) / (vertical ? scaleElement.clientHeight : scaleElement.clientWidth);
    var offsetValue = offset * scale;
    var value = rtl || vertical ? max - offsetValue : min + offsetValue;
    return snapValue(value, options);
}
exports.eventValue = eventValue;
/**
 * @hidden
 */
function isButton(element) {
    return element.className.indexOf('k-button-increase') >= 0 || element.className.indexOf('k-button-decrease') >= 0;
}
exports.isButton = isButton;
/**
 * @hidden
 */
function isDocumentAvailable() {
    return typeof document !== 'undefined';
}
exports.isDocumentAvailable = isDocumentAvailable;
