// @progress/kendo-angular-grid/dist/es/selection/selectable.directive.js
import { Directive, Input, ElementRef, Renderer2 } from '@angular/core';
import { SelectionService } from './selection.service';
/**
 * @hidden
 */
var SelectableDirective = (function () {
    function SelectableDirective(selectionService, el, renderer) {
        this.selectionService = selectionService;
        this.el = el;
        this.renderer = renderer;
        this.ignored = /^(a|input|textarea|button)$/i;
        if (this.selectionService.options.enabled) {
            this.destroyClick = this.renderer.listen(this.el.nativeElement, "mousedown", this.onMousedown.bind(this));
        }
    }
    SelectableDirective.prototype.ngOnDestroy = function () {
        if (this.destroyClick) {
            this.destroyClick();
        }
    };
    SelectableDirective.prototype.onMousedown = function (event) {
        if (!this.selectionService.options.enabled) {
            return;
        }
        if (this.shouldSelect(event.target.tagName) && !this.selectionService.options.checkboxOnly) {
            event.preventDefault();
            this.selectionService.handle(this.item, event);
        }
    };
    SelectableDirective.prototype.shouldSelect = function (tagName) {
        return !this.ignored.test(tagName);
    };
    return SelectableDirective;
}());
export { SelectableDirective };
SelectableDirective.decorators = [
    { type: Directive, args: [{
                selector: '[kendoGridSelectable]'
            },] },
];
/** @nocollapse */
SelectableDirective.ctorParameters = function () { return [
    { type: SelectionService, },
    { type: ElementRef, },
    { type: Renderer2, },
]; };
SelectableDirective.propDecorators = {
    'item': [{ type: Input },],
};
