// @progress/kendo-angular-inputs/dist/npm/common/events-outside-angular.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
/**
 * @hidden
 */
var EventsOutsideAngularDirective = (function () {
    function EventsOutsideAngularDirective(element, ngZone, renderer) {
        this.element = element;
        this.ngZone = ngZone;
        this.renderer = renderer;
        this.eventNames = [];
        this.domEventHandler = new core_1.EventEmitter();
    }
    EventsOutsideAngularDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (!this.element || !this.element.nativeElement) {
            return;
        }
        var names = [].concat(this.eventNames);
        this.subscriptions = [];
        this.emitEvent = this.emitEvent.bind(this);
        this.ngZone.runOutsideAngular(function () {
            for (var idx = 0; idx < names.length; idx++) {
                _this.subscriptions.push(_this.renderer.listen(_this.element.nativeElement, names[idx], _this.emitEvent));
            }
        });
    };
    EventsOutsideAngularDirective.prototype.ngOnDestroy = function () {
        if (this.subscriptions) {
            for (var idx = 0; idx < this.subscriptions.length; idx++) {
                this.subscriptions[idx]();
            }
            this.subscriptions = null;
        }
    };
    EventsOutsideAngularDirective.prototype.emitEvent = function (args) {
        this.domEventHandler.emit(args);
    };
    return EventsOutsideAngularDirective;
}());
EventsOutsideAngularDirective.decorators = [
    { type: core_1.Directive, args: [{
                selector: '[kendoEventsOutsideAngular]'
            },] },
];
/** @nocollapse */
EventsOutsideAngularDirective.ctorParameters = function () { return [
    { type: core_1.ElementRef, },
    { type: core_1.NgZone, },
    { type: core_1.Renderer2, },
]; };
EventsOutsideAngularDirective.propDecorators = {
    'eventNames': [{ type: core_1.Input },],
    'domEventHandler': [{ type: core_1.Output, args: ['kendoEventsOutsideAngular',] },],
};
exports.EventsOutsideAngularDirective = EventsOutsideAngularDirective;
